content/blog/2021/hidden_powers_of_bitcoin.md

```diff
@@ -0,0 +1,508 @@
+They make it easier to deal with Multisig or complicated key setups.
+Descriptors are super portable and can be easily used by any wallet to determine the list of all addresses that can be generated from the same.
+This feature creates a common stage for all Bitcoin apps and software.
+It's an abstract way to replace keys.
+
+According to Bitcoin Core, Output Descriptors are "a simple language which can be used to describe collections of output scripts".
+They bring in themselves, derivation paths, master xpub/xprv fingerprints and paths to generate addresses from.
+Let's understand this with an example of an Output Descriptor:
+
+```
+
+Descriptor: pkh([d34db33f/44'/0'/0']xpub6ERaJH[...]LJRcEL/1/*)#ml40v0wf
+            <1> <--------2---------><----------3---------><4> <---5--->
+
+Sections:
+1 - function type (here, P2PK)
+2 - derivation path of this key
+3 - xpub at m/44'/0'/0
+4 - path to deriving keys/addresses at
+5 - checksum for the descriptor
+```
+
+This was an example of a simple descriptor using one function but they can also be used in conjugation with multiple functions like `combo`, `multi` etc.
+The concept of descriptor came into existence in 2018 and since then, a lot of wallets have added support for descriptors.
+
+### Where it all comes together...
+
+In this section, we are going to make a descriptor-based wallet and derive addresses from `bitcoin-cli` and then use `bdk-cli` to confirm that the addresses generated for descriptor wallets are deterministic for a given descriptor.
+
+We will also try to create a vault miniscript policy and push funds to the vault with a lock time of 2 months. 
+During this time, we will try to break our vault and see our transactions failing.
+
+##### Tools and Armor
+
+- [docker](https://docs.docker.com/engine/install/)
+- [bdk-cli](https://github.com/bitcoindevkit/bdk-cli)
+- [miniscriptc](https://bitcoindevkit.org/bdk-cli/compiler/#installation)
+
+##### Setting Up
+
+We require `bitcoind` to run in `regtest` mode.
+The hard work has been already done in the `bitcoindevkit/electrs` docker image.
+The electrs functionality will also prove significant because `bdk` can use this local `electrs` box and send the transactions that we're gonna do for testing.
+
+```bash
+alias elstart='docker run --detach --rm -p 127.0.0.1:18443-18444:18443-18444/tcp -p 127.0.0.1:60401:60401/tcp --name electrs bitcoindevkit/electrs:0.5.0'
+alias elstop='docker kill electrs'
+alias ellogs='docker container logs electrs'
+alias elcli='docker exec -it electrs /root/bitcoin-cli -regtest -datadir=/root/.bitcoin $@'
+
+# Start Bitcoin Core
+elstart
+
+# Delete default wallet created by bitcoindevkit/electrs
+elcli unloadwallet "bdk-test"
+```
+
+Please note that `elcli` is an alias for `bitcoin-cli`.
+
+#### Keys and Generating Addresses
+
+Let us first generate an XPRV and create the wpkh wallet descriptor
```

Comment:
I tried `cargo install bdk-cli --features electrum --features compiler` but this didn't complete successfully.
It's ending in: 
```
error: no binaries are available for install using the selected features
```

Reply:
Ah that's because the new version 0.2.1 of bdk-cli hasn't been tagged yet. cc @notmandatory . Currently cargo is trying to install 0.2.0, which requires the `repl` feature and is not included by default.

This works with 0.2.0 `cargo install bdk-cli --features electrum --features compiler --features repl`.

But we should get the latest bdk-cli tagged before releasing this tutorial. Otherwise the instructions can break in future.

Reply:
Also curious, what version of bdk-cli did you use in the tutorial?